submissions/DmitryHniezdilov/task_js_oop/app.js

```diff
@@ -0,0 +1,296 @@
+    img: "images/char-boy.png",
+    initialBlock: {
+      x: 3,
+      y: 6,
+    },
+  },
+  ENEMIES_CONF: {
+    img: "images/enemy-bug.png",
+    initialRows: [2, 3, 4],
+    amountOfEnemies: [2, 2, 1],
+    speedIncrease: 0.2,
+    minSpeedOfEnemy: 50,
+    maxSpeedOfEnemy: 300,
+  },
+  STATE: {
+    currentLevel: 1,
+    speedMultiplicator: 1,
+  },
+  END_OF_GAME_WIN: "win",
+  END_OF_GAME_LOSE: "lose",
+  PRINT_MESSAGE_WIN: "You Win",
+  PRINT_MESSAGE_LOSE: "You Lose",
+  ELEMENT_POPUP_MESSAGE: document.querySelector(".js-popup"),
+  ELEMENT_INFO_LEVEL: document.querySelector(".js-info-level"),
+  ELEMENT_INFO_SPEED: document.querySelector(".js-info-speed"),
+  POPUP_MESSAGE_ACTIVE_CLASS: "is-active",
+};
+
+const Character = function (config, imgUrl) {
+  this.config = config;
+  this.sprite = imgUrl;
+};
+
+Character.prototype.endOfGame = function (result, config) {
+  const {
+    POPUP_MESSAGE_ACTIVE_CLASS,
+    ELEMENT_POPUP_MESSAGE,
+    ELEMENT_INFO_LEVEL,
+    ELEMENT_INFO_SPEED,
+    END_OF_GAME_WIN,
+    END_OF_GAME_LOSE,
+    PRINT_MESSAGE_WIN,
+    PRINT_MESSAGE_LOSE,
+    ENEMIES_CONF,
+    STATE,
+  } = config;
+
+  const updateMarkupInfo = (popupText) => {
+    ELEMENT_POPUP_MESSAGE.innerText = popupText;
+    ELEMENT_INFO_LEVEL.innerText = STATE.currentLevel;
+    ELEMENT_INFO_SPEED.innerText = `x${STATE.speedMultiplicator}`;
+    ELEMENT_POPUP_MESSAGE.classList.add(POPUP_MESSAGE_ACTIVE_CLASS);
+  };
+
+  isFreeze = true;
+  if (result === END_OF_GAME_WIN) {
+    STATE.currentLevel++;
+    STATE.speedMultiplicator = +(
+      STATE.speedMultiplicator + ENEMIES_CONF.speedIncrease
+    ).toFixed(1);
+    updateMarkupInfo(PRINT_MESSAGE_WIN);
+  } else if (result === END_OF_GAME_LOSE) {
+    STATE.currentLevel = 1;
+    STATE.speedMultiplicator = 1;
+    updateMarkupInfo(PRINT_MESSAGE_LOSE);
+  }
+
+  setTimeout(() => {
+    ELEMENT_POPUP_MESSAGE.classList.remove(POPUP_MESSAGE_ACTIVE_CLASS);
+    this.resetGame();
+    isFreeze = false;
+  }, 900);
+};
+
+const Enemy = function (x, y, minSpeed, maxSpeed, config) {
+  Character.call(this, config, config.ENEMIES_CONF.img);
+  this.ENEMIES_CONF = config.ENEMIES_CONF;
+  this.FIELD_WIDTH = config.FIELD_WIDTH;
+  this.BLOCK_WIDTH = config.BLOCK_WIDTH;
+  this.BLOCK_HEIGHT = config.BLOCK_HEIGHT;
+  this.x = x;
+  this.y = y;
+  this.speed = this.getRandomArbitrary(minSpeed, maxSpeed);
+};
+
+Enemy.prototype = Object.create(Character.prototype);
+
+Enemy.prototype.update = function (dt) {
+  if (!isFreeze) {
+    this.x += this.speed * dt;
+
+    if (this.x > this.FIELD_WIDTH) {
+      this.x = -this.BLOCK_WIDTH;
+    }
+
+    this.checkCollision();
+  }
+};
+
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.checkCollision = function () {
+  if (
+    player.y === this.y &&
+    player.x <= Math.floor(this.x) + this.BLOCK_WIDTH / 1.3 &&
+    player.x >= Math.floor(this.x) - this.BLOCK_WIDTH / 1.3
```

Comment:
The requirements address this case specifically.
Please find it, post here as a comment and fix code accordingly.
This is also applicable to other global variables with the exception of constants as described in the requirements.

Reply:
> Please find it, post here as a comment and fix code accordingly.

>"Requirements regarding Constants:
i. All numbers like block dimensions, initial locations are defined as constants.";
ii.  Every number that has a semantic purpose (like those listed above) should be defined as constants; think of how your code reads - the closer to plain English the better."

I missed adding the value as a constant in the config object.